### ISSUE TITLE
Docker for mac - Can't produce outside container

### ISSUE THREAD

**Issue by jmcarvalho** [wurstmeister/kafka-docker#255]
Hi. 
As I am running docker natively I used the _**docker.for.mac.localhost**_  as KAFKA_ADVERTISED_ HOST_NAME
Producing and consuming inside the container is working ok, however when I try to produce events outside the container setting **("bootstrap.servers", "localhost:9092")** I get a timeout exception probably due to the fact that the producer can't reach the cluster.
Is there any solution for that? 

docker- compose:
```
version: '2'
services:
  zookeeper:
    image: wurstmeister/zookeeper:latest
    ports:
      - "2181:2181"
    container_name: zookeeper
  kafka:
    image: wurstmeister/kafka:0.11.0.1
    ports:
      - "9092:9092"
    environment:
      KAFKA_ADVERTISED_HOST_NAME: docker.for.mac.localhost
      KAFKA_ZOOKEEPER_CONNECT: zookeeper:2181
      KAFKA_CREATE_TOPICS: "logs_topic:1:1"
    links:
      - zookeeper
    depends_on:
      - zookeeper
    container_name: kafka
```

**Comment by hwzhao** [comment]
Hi!
I have same issue. after setting hosts file to resolve the docker.for.mac.localhost as my machine ip, everything is ok.

**Comment by jmcarvalho** [comment]
which IP have you used in the hosts file? the same as localhost (127.0.0.1)?

**Comment by tcjennings** [comment]
`docker.for.mac.localhost` is for connecting to the Docker host from within a container. It is not the name to use in the Kafka advertised.listener configuration when your goal is to produce/consume from outside the Docker container network.

Instead, set KAFKA_ADVERTISED_HOST_NAME to the IP address of your Mac, which is effectively the same solution as @hwzhao above. Do not use localhost or 127.0.0.1.

**Comment by jmcarvalho** [comment]
@tcjennings I have kafka between two instances of logstash. First one ingests and the other one consume from kafka. By setting KAFKA_ADVERTISED_HOST_NAME to the IP address of my Mac I loose connection from the second logstash:
`ConsumerCoordinator - Cannot auto-commit offsets for group logstash since the coordinator is unknown`

**Comment by tcjennings** [comment]
Are your logstash producers and consumers all outside Docker?

See the docker-compose.yml from my pending PR #256:

https://github.com/tcjennings/kafka-docker/blob/master/docker-compose.yml

Try using your Mac's IP with that compose file. You should be able to consume and produce from outside Docker by giving your clients the bootstrap broker "${MAC_IP}:9094". 

The issue with single-listener configurations and outside-Docker clients is the the brokers also use the advertised.listeners to talk amongst themselves, which works (as long as there's nothing on the host to prevent it, like a firewall) but is suboptimal. I consider the best practice to be the two-listener approach described in the compose file linked above.

Don't use `container_name` (unnecessary because the service name is the thing) or `links` (deprecated and doesn't do anything)

**Comment by jmcarvalho** [comment]
Thanks for your help @tcjennings I will look into this compose.
By now I ended up following the @hwzhao suggestion, and after some other modifications things seem to work now, however as my IP is obviously changing whenever I connect to different networks, I would like to know if there is a way to have something to modify hosts file when it happens

**Comment by tcjennings** [comment]
That kind of host configuration question is not within the scope of this repository's audience. You might find some support on StackOverflow or a similar forum. Good luck.

**Comment by sscaling** [comment]
@jmcarvalho - is this resolved now?

**Comment by jmcarvalho** [comment]
@sscaling still the same I guess. Thereby, I removed the KAFKA_ADVERTISED_HOST_NAME from the compose file there's no need of it there. I only needed communication from host as I am processing with Kafka streams when kafka broker is inside a container. My workaround was to run the code inside the container. I think changing the hosts file is not a good practice. If your intention is to run docker natively on mac os you should either give a look into this: https://github.com/AlmirKadric-Published/docker-tuntap-osx or simply try to run your containers inside a docker machine.

**Comment by wwwreynolds** [comment]
I also struggled with this a while, but finally got something working by using the docker-file: https://github.com/simplesteph/kafka-stack-docker-compose/blob/master/zk-single-kafka-single.yml, which uses confluent's kafka docker impl.

With the latest docker for mac (and if you don't want to use docker-machine), I think you need to add something like this to your /etc/hosts file for your mac:
```
127.0.0.1       kafka
127.0.0.1       zookeeper
```
Then, in your docker-compose, use these environment settings for kafka:
```
      KAFKA_ADVERTISED_LISTENERS: "PLAINTEXT://kafka:9092"
      KAFKA_ZOOKEEPER_CONNECT: "zookeeper:2181"
```

So, inside the docker container 'kafka' resolves to the docker network's IP address, but outside the container 'kafka' resolves to localhost.

I tried using these settings with the wurstmeister image but could never get it to work, even with lot's of tweaking - but I think it's close. I may get back to this later and hopefully give you a better answer.

**Comment by sscaling** [comment]
@wwwreynolds - this is next in my list of things to fix (https://github.com/wurstmeister/kafka-docker/issues/221). Just awaiting another PR to be merged first.

To work around this, you'd need to specify `KAFKA_ADVERTISED_HOST_NAME: kafka` which will configure the `advertised.listeners` property as required.

**Comment by sscaling** [comment]
This should be fixed by #286

**Comment by sscaling** [comment]
@jmcarvalho - it sounds like some sort of configuration issue. Please make sure you understand the requirements in https://github.com/wurstmeister/kafka-docker/wiki/Connectivity. That will hopefully help you configure the image as required.

Closing as #221 tracks the work required for this.
